proposals/p2029.md

```diff
@@ -0,0 +1,103 @@
+# C++ Interop: Reserved identifiers
+
+<!--
+Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+Exceptions. See /LICENSE for license information.
+SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+-->
+
+[Pull request](https://github.com/carbon-language/carbon-lang/pull/2029)
+
+<!-- toc -->
+
+## Table of contents
+
+-   [Problem](#problem)
+-   [Background](#background)
+-   [Proposal](#proposal)
+-   [Details](#details)
+-   [Rationale](#rationale)
+-   [Alternatives considered](#alternatives-considered)
+
+<!-- tocstop -->
+
+## Problem
+
+C++ reserves certain sets of identifiers for implementation-specific use. Carbon
+must ensure it respects these reserved identifiers to enable seamless and
+error-free interoperability with C++. Equally, it would be prudent to define a
+set of identifiers which are reserved for internal Carbon implementation usage.
+
+## Background
+
+The full set of C++ reserved identifiers is listed
+[here](https://en.cppreference.com/w/cpp/language/identifiers). This proposal is
+solely considering the set of user-definable identifiers which are considered
+names in C++:
+
+-   Identifiers with a double underscore anywhere are reserved;
+-   Identifiers that begin with an underscore followed by an uppercase letter
+    are reserved;
+-   Identifiers that begin with an underscore are reserved in the global
+    namespace.
+
+Considerations for reserved keywords are discussed in
+[p0093](https://github.com/carbon-language/carbon-lang/pull/93).
+
+## Proposal
+
+Carbon reserves the following identifiers in the specified contexts:
+
+-   Identifiers with a double underscore anywhere are reserved;
+-   Identifiers that begin with an underscore followed by an uppercase letter
+    are reserved.
+
+Carbon toolchains are expected to emit an error when a reserved identifier is
+encountered. This will remove the risk of undefined / unexpected runtime
+behaviour, or name resolution clashes when linking with C++ libraries.
+
+## Details
+
+This proposal makes use of three assumptions:
+
+1. A Carbon package exporting a C++ header will be scoped in a C++ namespace
+   following the package name;
```

Comment:
To explain the behavior I expect:

By default, some (but not all) Carbon names will be exported to a matching C++ namespace. However, this won't include everything (e.g., some Carbon features will preclude availability in C++). Also, there'll probably be some capacity for renaming the exported symbol. Off the top of my head, I have two reasons, both based on migration of C++ code: (1) there may be an API in C++'s global namespace that's being replaced; (2) Carbon's naming and package idioms may not match how C++ code was written.

So what you're describing here is a default behavior, but not the only behavior.

Reply:
Is there any documentation on what the non-default behaviour could look like? Or does this proposal need to wait for futher interop design / evolution work?

Reply:
[This doc](https://docs.google.com/document/d/1rPJ5GQnd-TLyFULUNNwfl5Mf2fIpX8mR7qcI4SjpQXw/edit?resourcekey=0-A7I-NiJbomij1b0gUW26PQ) has a lot of thought on this, but note that's pretty old and the related PR was deferred.